packages/lib/src/verify.ts

```diff
@@ -0,0 +1,245 @@
+ * @param pathToCircuits - The path to the verification key. Only needed for server side verification
+ * @returns - A boolean indicating whether or not all of the proofs are valid
+ */
+export const batchVerifyMembership = async (
+  proofs: EcdsaMembershipProof[],
+  pubKeys: WeierstrassPoint[],
+  nullifierRandomness: bigint = BigInt(0),
+  pathToCircuits: string | undefined = undefined
+): Promise<boolean> => {
+  if (isNode() && pathToCircuits === undefined) {
+    throw new Error(
+      "Path to circuits must be provided for server side verification!"
+    );
+  }
+
+  console.time("Batch Membership Proof Verification");
+  console.time("Batch Merkle Root Computation");
+  const edwardsPubKeys = pubKeys.map((pubKey) => pubKey.toEdwards());
+  const computedMerkleRoot = await computeMerkleRoot(edwardsPubKeys);
+  console.timeEnd("Batch Merkle Root Computation");
+
+  console.time("Fetching Verification Key");
+  const vKey = isNode()
+    ? await getVerificationKeyFromFile(pathToCircuits!)
+    : await getVerificationKeyFromUrl();
+  console.timeEnd("Fetching Verification Key");
+
+  const verified = await Promise.all(
+    proofs.map(async (proof, i) => {
+      console.time(`Membership Proof Verification: ${i}`);
+      console.time(`Merkle Root Verification: ${i}`);
+      const publicSignals = proof.zkp.publicSignals;
+      const merkleRoot = BigInt(publicSignals[1]);
+      if (computedMerkleRoot !== merkleRoot) {
+        return false;
+      }
+      console.timeEnd(`Merkle Root Verification: ${i}`);
+
+      console.time(`T and U Verification: ${i}`);
+      const [Tx, Ty, Ux, Uy] = publicSignals.slice(2, 6).map(BigInt);
+      const T = new EdwardsPoint(Tx, Ty);
+      const U = new EdwardsPoint(Ux, Uy);
+      const { R, msgHash } = proof;
+      const { T: computedT, U: computedU } = await recoverTUFromProof(
+        R,
+        msgHash
+      );
+      if (!computedT.equals(T) || !computedU.equals(U)) {
+        return false;
+      }
+      console.timeEnd(`T and U Verification: ${i}`);
+
+      console.time(`Nullifier Verification: ${i}`);
+      const proofNullifierRandomness = BigInt(publicSignals[6]);
+      if (proofNullifierRandomness !== nullifierRandomness) {
+        return false;
+      }
+      console.timeEnd(`Nullifier Verification: ${i}`);
+
+      console.time(`ZK Proof Verification: ${i}`);
+      const verified = await verifyMembershipZKP(vKey, proof.zkp);
+      console.timeEnd(`ZK Proof Verification: ${i}`);
+      console.timeEnd(`Membership Proof Verification: ${i}`);
+
+      return verified;
+    })
+  );
+
+  // snarkjs will not terminate this object automatically
+  // We should do so after all proving/verification is finished for caching purposes
+  // See: https://github.com/iden3/snarkjs/issues/152
+  // @ts-ignore
+  await globalThis.curve_bn128.terminate();
+  console.timeEnd("Batch Membership Proof Verification");
+
+  // Could rewrite this to short circuit if any are false verifications,
```

Comment:
Perhaps have this be a flag that can be passed in? Also tbh I can't think of any cases where it is useful to know which ones have failed except for like test cases

Reply:
Yeah I just removed the comment